File: 4.Median_of_Two_Sorted_Arrays/binarysol.py
class Solution:
    def findMedianSortedArrays(self,nums1,nums2)->float:
        n1=len(nums1);
        n2 = len(nums2);
        if n1>n2:
            nums1,nums2 = nums2,nums1
        n = n1 +n2
        left = (n+1)//2
        low,high = 0,n1

        while low<=high:
            mid1 = (high+low)//2
            mid2 = left-mid1

            l1 = float('-inf')
            l2 = float('-inf')
            r1 = float('inf')
            r2 = float('inf')

            if mid1<n1:
                r1=nums1[mid1]
            if mid2<n2:
                r2=nums2[mid2]
            if mid1-1>=0:
                l1=nums1[mid1-1]
            if mid2-1>=0:
                l2=nums2[mid2-1]
            
            if l1<=r2 and l2<=r1:
                if n%2 == 1:
                    return max(l1,l2)
                else:
                    return (max(l1,l2) + min(r1,r2))/2
            elif l1>r2:
                high = mid1-1
            else:
                low = mid1+1
        return 0

if __name__== "__main__":
    s= Solution()
    print(s.findMedianSortedArrays([1,2],[3,4,5])) # 2.0

File: 4.Median_of_Two_Sorted_Arrays/binarysol.java
import java.util.Arrays;

class Solution{
    public double findMedianSortedArrays(int[] nums1, int[] nums2){
        int n1 = nums1.length;
        int n2 = nums2.length;
        if(n1 > n2){
            return findMedianSortedArrays(nums2, nums1);
        }
        int low = 0;
        int high = n1;
        int n = n1+n2;
        int left = (n+1)/2;

        while (low<=high){
            int mid1 = (high+low)/2;
            int mid2 = left - mid1;

            int l1 = (mid1 == 0) ? Integer.MIN_VALUE : nums1[mid1-1];
            int l2 = (mid2 == 0) ? Integer.MIN_VALUE : nums2[mid2-1];
            int r1 = (mid1 == n1) ? Integer.MAX_VALUE : nums1[mid1];
            int r2 = (mid2 == n2) ? Integer.MAX_VALUE : nums2[mid2];


            if(l1<=r2 && l2<=r1){
                if(n%2 == 0){
                    return (Math.max(l1, l2) + Math.min(r1, r2))/2.0;
                }else{
                    return Math.max(l1, l2);
                }
            }else if(l1>r2){
                high = mid1-1;
            }else{
                low = mid1+1;
            }
        }
    }
}

class binarysol{
    public static void main(String[] args){
        Solution s = new Solution();
        int[] nums1 = {1, 3};
        int[] nums2 = {2};
        System.out.println(s.findMedianSortedArrays(nums1, nums2));
    }
}